src/compile/layer.ts

```diff
@@ -89,39 +87,62 @@ export class LayerModel extends Model {
-          if (!newScales) {
+          let childScales: ScaleComponents = child.component.scale[channel];
+          if (!childScales) {
+            // the child does not have any scales so we have nothing to merge
             return;
           }
 
-          const existingScales = scaleComponent[channel];
-          if (existingScales && newScales[0].type === ScaleType.ORDINAL) {
-            // Ordinal scales are unioned by combining the domain of the first scale. Other scales are appended.
-            const existingDomain = existingScales[0].domain.fields ? existingScales[0].domain : { fields: [existingScales[0].domain] };
-            newScales[0].domain = {
-              // assumes that the new scale domain is not a list
-              fields: existingDomain.fields.concat([newScales[0].domain])
-            };
-            if (existingScales.length > 1) {
-              delete existingScales[0];
-              scaleComponent[channel] = newScales.concat(existingScales);
+          const modelScales: ScaleComponents = scaleComponent[channel];
+          if (modelScales && modelScales.main) {
+            // Scales are unioned by combining the domain of the first scale (the main scale).
+            // Other scales that are used for ordinal legends are appended.
+            const modelDomain = modelScales.main.domain;
+            const childDomain = childScales.main.domain;
+
+            if (isArray(modelDomain)) {
+              if (isArray(childScales.main.domain)) {
+                modelScales.main.domain = modelDomain.concat(childDomain);
+              } else {
+                model.addWarning('scales cannot be unioned with explicit domain');
+              }
             } else {
-              scaleComponent[channel] = newScales;
+              const unionedFields = isUnionedDomain(modelDomain) ? modelDomain.fields : [modelDomain] as VgDataRef[];
+
+              if (isArray(childDomain)) {
+                model.addWarning('scales cannot be unioned with explicit domain');
+              }
+
+              let fields = isDataRefDomain(childDomain) ? unionedFields.concat([childDomain]) :
+                // if the domain is itself a union domain, concat
+                isUnionedDomain(childDomain) ? unionedFields.concat(childDomain.fields) :
+                  // we have to ignore explicit data domains for now because vega does not support unioning them
+                  unionedFields;
+              fields = unique(fields, hash);
+              // TODO: if all domains use the same data, we can merge them
+              if (fields.length > 1) {
+                modelScales.main.domain = { fields: fields };
+              } else {
+                modelScales.main.domain = fields[0];
+              }
             }
+
+            // create color legend and color legend bin scales if we don't have them yet
+            modelScales.colorLegend = modelScales.colorLegend ? modelScales.colorLegend : childScales.colorLegend;
```

Comment:
Why does domain of the color field in `colorLegend` never get merged?   

It seems like there are 3 cases:

- **ordinal**:   this is kinda dumb to simply merge two ordinal fields together.  It seems like the right thing to do is to provide a custom domain for the union domain.   (Another VL 1.0 bug: I also notice that the ordinal color scale would behave incorrect if we provide explicit custom domain)
- **bin**:  This is tricky.   It seems like the field should be unioned before binning, I haven't tracked back the code to see how this should work.  
- **timeUnit**:  Similar to bin. 

Maybe we should file an issue for this and fix them later as it should not be show stopping for our paper's demo?